devel/ci/cico.pipeline

```diff
@@ -0,0 +1,169 @@
+    // Synchronize the given path to the Duffy node from the control host.
+    sh 'rsync -e "ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -l root " -Ha --include=' +  rsyncpath + " ./ ${DUFFY_NODE}.ci.centos.org:~/payload/"
+}
+
+
+/**
+ * Get a Duffy node and set the DUFFY_NODE and SSID environment variables.
+ */
+def get_duffy_node = {
+    // Get a duffy node and set the DUFFY_NODE and SSID environment variables.
+    duffy_rtn=sh(
+        script: 'cico --debug node get -f value -c hostname -c comment',
+        returnStdout: true
+        ).trim().tokenize(' ')
+    env.DUFFY_NODE=duffy_rtn[0]
+    env.SSID=duffy_rtn[1]
+}
+
+
+/**
+ * Install test dependencies on the Duffy node.
+ */
+def configure_node = {
+    // Install packages we need on the Duffy node.
+    onmyduffynode 'yum -y install epel-release'
+    onmyduffynode 'yum install -y docker python34-click rsync'
+    onmyduffynode 'systemctl start docker'
+}
+
+
+/**
+ * Run the given function up to 5 times (if it fails), with 60 seconds of sleeping in between tries.
+ *
+ * @param f The function to retry.
+ */
+def retry_with_sleep = { f ->
+    try {
+        f()
+    } catch (error) {
+        retry(4) {
+            sleep(60)
+            f()
+        }
+    }
+}
+
+
+/**
+ * Run bodhi-ci on the Duffy node for the given arguments.
+ *
+ * @param release The release to pass to bodhi-ci's -r flag.
+ * @param command The subcommand of bodhi-ci to use.
+ * @param context This parameter is appended to the release and a dash to form the GitHub context.
+ * @param args Any extra arguments to be passed to the command.
+ */
+def bodhi_ci = { String release, String command, String context, String args ->
+    githubNotify context: release + '-' + context, status: 'PENDING'
+
+    try {
+        stage(release + ' ' + command) {
+            timeout(time: 1, unit: 'HOURS') {
+                onmyduffynode 'cd payload && ./devel/ci/bodhi-ci ' + command + ' -r ' + release + ' ' + args
+            }
+        }
+        githubNotify context: release + '-' + context, status: 'SUCCESS'
+    } catch(error) {
+        githubNotify context: release + '-' + context, status: 'FAILURE'
+    }
+}
+
+
+/**
+ * Build and test the given release.
+ *
+ * @param release The release to test.
+ */
+def test_release = { String release ->
+    bodhi_ci(release, 'build', 'build', '')
+
+    parallel(
+        docs: {bodhi_ci(release, 'docs', 'docs', '--no-build')},
+        flake8: {bodhi_ci(release, 'flake8', 'flake8', '--no-build')},
+        pydocstyle: {bodhi_ci(release, 'pydocstyle', 'pydocstyle', '--no-build')},
+        unitpy2: {bodhi_ci(release, 'unit', 'python2-unit', '--no-build -a -p 2')},
+        unitpy3: {bodhi_ci(release, 'unit', 'python3-unit', '--no-build -a -p 3')}
+    )
+}
+
+
+node('bodhi') {
+    checkout scm
+
+    stage('Allocate Duffy node') {
+        env.CICO_API_KEY = readFile("${env.HOME}/duffy.key").trim()
+        retry_with_sleep(get_duffy_node)
```

Comment:
[nonblocking] instead of doing the retry in groovy, I'd try the `--retry-interval` and `--retry-count` options for cicoclient. That would help us respect limits once duffy passes rate-limit info back to the client.

Reply:
@bstinsonmhk I'm applying the following patch to address that idea and also because I found a couple duplicate comments (duplicates because they were faux-docblocks and now there are real docblocks that say the same thing):

```
diff --git a/devel/ci/cico.pipeline b/devel/ci/cico.pipeline
index e9cd68c6..9dbd90bd 100644
--- a/devel/ci/cico.pipeline
+++ b/devel/ci/cico.pipeline
@@ -26,7 +26,6 @@ def onmyduffynode(String script) {
  * @param rsyncpath The path to be rsync'd back to the control host.
  */
 def syncfromduffynode(rsyncpath) {
-    // Synchronize the given path from the Duffy node back to the control host.
     sh 'rsync -e "ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -l root " -Ha --include=' + " ${DUFFY_NODE}.ci.centos.org:~/payload/" + rsyncpath + " ./"
 }
 
@@ -37,30 +36,14 @@ def syncfromduffynode(rsyncpath) {
  * @param rsyncpath The path to be rsync'd to the Duffy node.
  */
 def synctoduffynode(rsyncpath) {
-    // Synchronize the given path to the Duffy node from the control host.
     sh 'rsync -e "ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -l root " -Ha --include=' +  rsyncpath + " ./ ${DUFFY_NODE}.ci.centos.org:~/payload/"
 }
 
 
-/**
- * Get a Duffy node and set the DUFFY_NODE and SSID environment variables.
- */
-def get_duffy_node = {
-    // Get a duffy node and set the DUFFY_NODE and SSID environment variables.
-    duffy_rtn=sh(
-        script: 'cico --debug node get -f value -c hostname -c comment',
-        returnStdout: true
-        ).trim().tokenize(' ')
-    env.DUFFY_NODE=duffy_rtn[0]
-    env.SSID=duffy_rtn[1]
-}
-
-
 /**
  * Install test dependencies on the Duffy node.
  */
 def configure_node = {
-    // Install packages we need on the Duffy node.
     onmyduffynode 'yum -y install epel-release'
     onmyduffynode 'yum install -y docker python34-click rsync'
     onmyduffynode 'systemctl start docker'
@@ -131,7 +114,13 @@ node('bodhi') {
 
     stage('Allocate Duffy node') {
         env.CICO_API_KEY = readFile("${env.HOME}/duffy.key").trim()
-        retry_with_sleep(get_duffy_node)
+        // Get a duffy node and set the DUFFY_NODE and SSID environment variables.
+        duffy_rtn=sh(
+            script: 'cico --debug node get -f value -c hostname -c comment --retry-count 4 --retry-interval 60',
+            returnStdout: true
+            ).trim().tokenize(' ')
+        env.DUFFY_NODE=duffy_rtn[0]
+        env.SSID=duffy_rtn[1]
     }
 
     try {
```

Reply:
I feel confident approving this before CI finishes because the CI run for (https://ci.centos.org/job/bodhi-pipeline/job/PR-2609/41/consoleFull) successfully got a node from duffy at the start

:+1: